### PR TITLE
FEMS Requests combined to avoid timeouts

### PR DESCRIPTION
**Pull request by sthhs**
My FEMS (Fenecon Home 10, Version 2022 and later) is very sensitive with answering requests at the REST/JSON interface. Too many requests lead to very long answering times (30 s or more) and therefore to timeouts in the OpenWB. In the OpenEMS community you can find that there is indeed a limitation for not overloading the FEMS. The impact on the OpenWB is shown on the attached picture. 
Timeouts
With my modifications (combining of the requests in inverter.py, bat.py, counter.py) and increasing the timeout value in counter.py to 6 almost all of the peaks (occurences of timeouts that lead to the wrong calculation of the consumption) disappear. They aren't gone totally, but reduced to a very small quantity.

### CODE REVIEW

**Comment by sthhs** [comment]
Here are two more pictures.
Without the modification:
before
and with:
after

File: packages/modules/devices/fems/bat.py [hunk 1]
```diff
@@ -30,20 +30,26 @@ def update(self, session: Session) -> None:
...
+       #     timeout=2).json()
+       # for singleValue in response:
+       #     address = singleValue["address"]
+       #     if (address == "_sum/GridActivePower"):
+       #         grid = scale_metric(singleValue['value'], singleValue.get('unit'), 'W')
+       #     elif address == "_sum/ProductionActivePower":
+       #         pv = scale_metric(singleValue['value'], singleValue.get('unit'), 'W')
+       #     elif address == "_sum/ConsumptionActivePower":
+       #         haus = scale_metric(singleValue['value'], singleValue.get('unit'), 'W')
+
```
Comment (LKuemmel) [inline packages/modules/devices/fems/bat.py:52]:
```suggestion

```

Comment (LKuemmel) [inline packages/modules/devices/fems/bat.py:52]:
```suggestion

```
Please remove, GitHub is already tracking changes

File: packages/modules/devices/fems/counter.py [hunk 2]
```diff
@@ -44,18 +44,21 @@ def update(self, session: Session) -> None:
...
+            # response = session.get(
+            #     'http://'+self.ip_address+':8084/rest/channel/_sum/Grid.+ActiveEnergy',
+            #     timeout=1).json()
+
+            # for singleValue in response:
+            #     address = singleValue['address']
+            #     if (address == '_sum/GridBuyActiveEnergy'):
+            #         imported = scale_metric(singleValue['value'], singleValue.get('unit'), 'Wh')
+            #     elif (address == '_sum/GridSellActiveEnergy'):
+            #         exported = scale_metric(singleValue['value'], singleValue.get('unit'), 'Wh')
```
Comment (LKuemmel) [inline packages/modules/devices/fems/counter.py:61]:
```suggestion
```

File: packages/modules/devices/fems/inverter.py [hunk 3]
```diff
@@ -16,14 +16,22 @@ def __init__(self, ip_address: str, component_config: FemsInverterSetup) -> None
...
-            'http://'+self.ip_address+':8084/rest/channel/_sum/ProductionActiveEnergy',
-            timeout=2).json()
-        exported = scale_metric(response["value"], response.get("unit"), 'Wh')
+
+        # power = scale_metric(response["value"], response.get("unit"), 'W') * -1
+
+        # response = session.get(
+        #     'http://'+self.ip_address+':8084/rest/channel/_sum/ProductionActiveEnergy',
+        #     timeout=2).json()
+        # exported = scale_metric(response["value"], response.get("unit"), 'Wh')
```
Comment (LKuemmel) [inline packages/modules/devices/fems/inverter.py:34]:
```suggestion
```

**Review by LKuemmel — CHANGES_REQUESTED** [review]

**Comment by sthhs** [comment]
Changes completed

**Review by LKuemmel — APPROVED** [review]

**Comment by LKuemmel** [comment]
Bitte noch die Flake8-Änderungen vornehmen, damit der Test durchläuft.

**Review by LKuemmel — CHANGES_REQUESTED** [review]
Bitte noch die Flake8-Änderungen vornehmen, damit der Test durchläuft.

**Comment by sthhs** [comment]
Danke für den Hinweis, die Korrekturen sind gemacht.

**Comment by sthhs** [comment]
Die Antworten stehen jeweils unter der Abfrage:

**inverter.py:**
http://'+self.ip_address+':8084/rest/channel/_sum/(ProductionActivePower|ProductionActiveEnergy)

`[
    {
        "address": "_sum/ProductionActiveEnergy",
        "type": "LONG",
        "accessMode": "RO",
        "text": "",
        "unit": "Wh_Σ",
        "value": 18988374
    },
    {
        "address": "_sum/ProductionActivePower",
        "type": "INTEGER",
        "accessMode": "RO",
        "text": "Total production; always positive",
        "unit": "W",
        "value": -9
    }
]`

**counter.py:**
'http://' + self.ip_address +':8084/rest/channel/(meter0|_sum)/' + '(ActivePower.*|VoltageL.|Frequency|Grid.+ActiveEnergy)'

`[
    {
        "address": "meter0/ActivePowerL3",
        "type": "INTEGER",
        "accessMode": "RO",
        "text": "",
        "unit": "W",
        "value": 6141
    },
    {
        "address": "meter0/ActivePower",
        "type": "INTEGER",
        "accessMode": "RO",
        "text": "",
        "unit": "W",
        "value": 17189
    },
    {
        "address": "meter0/ActivePowerL1",
        "type": "INTEGER",
        "accessMode": "RO",
        "text": "",
        "unit": "W",
        "value": 5575
    },
    {
        "address": "meter0/ActivePowerL2",
        "type": "INTEGER",
        "accessMode": "RO",
        "text": "",
        "unit": "W",
        "value": 5473
    },
    {
        "address": "meter0/VoltageL1",
        "type": "INTEGER",
        "accessMode": "RO",
        "text": "",
        "unit": "mV",
        "value": 229500
    },
    {
        "address": "meter0/VoltageL2",
        "type": "INTEGER",
        "accessMode": "RO",
        "text": "",
        "unit": "mV",
        "value": 229300
    },
    {
        "address": "meter0/VoltageL3",
        "type": "INTEGER",
        "accessMode": "RO",
        "text": "",
        "unit": "mV",
        "value": 229300
    },
    {
        "address": "meter0/Frequency",
        "type": "INTEGER",
        "accessMode": "RO",
        "text": "",
        "unit": "mHz",
        "value": 50000
    },
    {
        "address": "_sum/GridSellActiveEnergy",
        "type": "LONG",
        "accessMode": "RO",
        "text": "",
        "unit": "Wh_Σ",
        "value": 11752059
    },
    {
        "address": "_sum/GridBuyActiveEnergy",
        "type": "LONG",
        "accessMode": "RO",
        "text": "",
        "unit": "Wh_Σ",
        "value": 1088853
    }
]
`
**bat.py:**
"http://" + self.ip_address + ":8084/rest/channel/(ess0|_sum)/" + "(Soc|DcChargeEnergy|DcDischargeEnergy|GridActivePower|ProductionActivePower|ConsumptionActivePower)"

`[
    {
        "address": "_sum/ConsumptionActivePower",
        "type": "INTEGER",
        "accessMode": "RO",
        "text": "",
        "unit": "W",
        "value": 17183
    },
    {
        "address": "_sum/ProductionActivePower",
        "type": "INTEGER",
        "accessMode": "RO",
        "text": "Total production; always positive",
        "unit": "W",
        "value": -9
    },
    {
        "address": "_sum/GridActivePower",
        "type": "INTEGER",
        "accessMode": "RO",
        "text": "Grid exchange power. Negative values for sell-to-grid; positive for buy-from-grid",
        "unit": "W",
        "value": 17164
    },
    {
        "address": "ess0/Soc",
        "type": "INTEGER",
        "accessMode": "RO",
        "text": "",
        "unit": "%",
        "value": 6
    },
    {
        "address": "ess0/DcChargeEnergy",
        "type": "LONG",
        "accessMode": "RO",
        "text": "",
        "unit": "Wh_Σ",
        "value": 3013394
    },
    {
        "address": "ess0/DcDischargeEnergy",
        "type": "LONG",
        "accessMode": "RO",
        "text": "",
        "unit": "Wh_Σ",
        "value": 3202108
    }
]`

**Comment by LKuemmel** [comment]
Sorry, ich habe mich lange nicht gemeldet. Ich habe die Tests erstellt bzw angepasst und war so frei, sie in Deinen Branch zu pushen. Danke für deine Beispieldaten!

**Review by LKuemmel — APPROVED** [review]

File: packages/modules/devices/fems/bat.py [hunk 4]
```diff
@@ -20,7 +20,8 @@ def update(self, session: Session) -> None:
         else:
             data = "ess2"
         response = session.get(
-            "http://" + self.ip_address + ":8084/rest/channel/"+data+"/(Soc|DcChargeEnergy|DcDischargeEnergy)",
+            "http://" + self.ip_address + ":8084/rest/channel/(" + data + "|_sum)/" +
```
Comment (dj3mu) [inline packages/modules/devices/fems/bat.py:23]: Sorry, but this change breaks queries on older FEMS versions.

At least on my FEMS v 2022.12.5 a query with Regex in any "segment" other than the last one is NOT supported. FEMS just returns:

`requests.exceptions.HTTPError: 404 Client Error: Not Found for url: http://<IP>:8084/rest/channel/(meter0%7C_sum)/(ActivePower.*%7CVoltageL.%7CFrequency%7CGrid.+ActiveEnergy)`

I also see the optimization done here. But I might not be the only one with older version. Actually I have no control on if and when Fenecon updates my FEMS.

So I have no other idea but reverting this!

**Pull request merged**